### PR TITLE
Distance and losses refactor

### LINKED ISSUE #92: Support XMB
https://github.com/bnu-wangxun/Deep_Metric

### PR DESCRIPTION
**Pull request by chjort**
This PR proposes to have `Distance` take two embedding tensors as input, and the `losses` module to support computing losses for asymmetric and non-square distance matrices.

These changes will accomodate new features such as [Cross-batch-memory (XBM)](https://arxiv.org/pdf/1912.06798.pdf) and will make it easier to close #92 in a new future pull request. Furthermore, these changes will make `Distance` and `losses` support a wider range of usecases, without much added complexity.

The `call` method of `Distance` will now have the following signature:
```python
def call(self, query_embeddings: FloatTensor, key_embeddings: FloatTensor) -> FloatTensor:
    """Compute distance(query_embeddings, key_embeddings)"""
```
instead of:

```python
def call(self, embeddings: FloatTensor) -> FloatTensor:
    """Compute distance(embeddings, embeddings)"""
```

Mathematically, it is quite intuitive and natural for a distance function to take two inputs, so confusion should be
minimal for users when they try to implement custom distances using `Distance`. It will also follow a similar pattern as [Distances](https://kevinmusgrave.github.io/pytorch-metric-learning/distances/#basedistance) in [Pytorch Metric Learning](https://kevinmusgrave.github.io/pytorch-metric-learning)

Right now the self-supervised contrastive losses implemented in `simclr.py`, `barlow.py` and `simsiam.py`
can not make use of any `Distance` implementation to compare the embeddings of two different views.
Since `Distance` is highly integrated into the supervised similarity losses, it would be natural for them to also
be integratable into the self-supervised losses.
For example by using `CosineDistance` in `simsiam.py` instead of re-implementing it on [line 17](https://github.com/tensorflow/similarity/blob/a671b5289b96016f453ef7b1f216d4db009cdc0c/tensorflow_similarity/losses/simsiam.py#L17).

The loss functions in the `losses` module are changed to have the following signature:
```python
def triplet_loss(query_labels: IntTensor,
                 query_embeddings: FloatTensor,
                 key_labels: IntTensor,
                 key_embeddings: FloatTensor,
                 distance: Callable,
                 remove_diagonal: bool = True,
                 #... loss hyperparameters
                 ) -> Any:
```
instead of:

```python
def triplet_loss(labels: IntTensor,
                 embeddings: FloatTensor,
                 distance: Callable,
                 #... loss hyperparameters
                 ) -> Any:
```

This makes it possible to compute the loss between two different embedding tensors each with their own labels.
This will be necessary to support XBM.

The additional input arguments added to `Distance` and to loss functions are required positional arguments.
They could be optional, but from the discussion in #92 it was decided that making them required is less confusing. 

NOTE: This exact PR does not close #92

Please let me know what you think, and any suggested changes are more than welcome!

### CODE REVIEW

**Comment by owenvallis** [comment]
This looks great! Thanks for including the additional test coverage.
